core/string/ustring.cpp

```diff
@@ -1644,6 +1644,33 @@ String String::hex_encode_buffer(const uint8_t *p_buffer, int p_len) {
 	return ret;
 }
 
+Vector<uint8_t> String::hex_decode() const {
+	ERR_FAIL_COND_V_MSG(length() % 2 != 0, Vector<uint8_t>(), "Hexadecimal string of uneven length.");
+
+#define HEX_TO_BYTE(c, o)                                                                                                                     \
```

Comment:
I don't think a macro is needed here. The code is repeated only once and will definitely not be repeated one more time. In addition, macros of this kind do not have to be universal and function-like. I think there is no need for additional variables and pointers.

Your macro expands into calling `operator[]()` multiple times instead of `c`. I don't know, maybe the compiler optimizes this. But why should we rely on it when we can use simpler code? If you want a macro, then I would do it like this:

```c++
#define HEX_TO_BYTE(m_k)                                                                                 \
    if (is_digit(c)) {                                                                                   \
        n += m_k * (c - '0');                                                                            \
    } else if (c >= 'a' && c <= 'f') {                                                                   \
        n += m_k * ((c - 'a') + 10);                                                                     \
    } else if (c >= 'A' && c <= 'F') {                                                                   \
        n += m_k * ((c - 'a') + 10);                                                                     \
    } else {                                                                                             \
        ERR_FAIL_V_MSG(Vector<uint8_t>(), "Invalid hexadecimal notation character \"" + chr(c) + "\"."); \
    }
```

But again, I don't see any problem with code duplication here.

Also note that we use `#undef` after the macro is no longer needed.

Reply:
i dont like duplicating the text.